### ISSUE TITLE
LFS compression format

### ISSUE THREAD

**Issue by emoose** [nipkownix/re4_tweaks#20]
I saw Albert mentioned on twitter a few weeks ago that part of reason for RE4HD's large filesize is due to .lfs not being able to be re-packed atm: https://twitter.com/re4hdproject/status/1454153402787708931

Might be nice to look into that, if anyone knows about any existing tools for it please let me know, only thing I've found so far is a quickbms script, but that does give some useful info:
```
comtype xmemdecompress
get name basename
log MEMORY_FILE 0 0 
goto 0x10
get chunks long
append
for i = 0 < chunks
get zsize short
get size short
if size = 0 
    math size = 0x10000
endif
get offset long
math offset + 0x13
clog MEMORY_FILE offset zsize size
next i
append
get size asize MEMORY_FILE
log name 0x00 size MEMORY_FILE
```

Seems they use the XMemDecompress xbox360-api for compression, pretty sure I've seen tools for compressing that around before, maybe just a matter of adding on whatever weird header lfs files start with.

E: ah okay format is pretty simple, hardest bit is probably the weird way they store offsets to each compressed chunk (seems to be aligned to each 0x14 bytes or something stupid)
Unfortunately couldn't find any working XMemCompress code on the net, but luckily I had some old code from messing about on x360 stuff that does seem to work, will try putting something together soon.

**Comment by emoose** [comment]
Aha sweet, managed to get something working: [RE4LFS-0.1.zip](https://github.com/nipkownix/re4_tweaks/files/7550783/RE4LFS-0.1.zip) (x64 only, src included)

Seems to match the compression with the original files fine, but I've only really tested it with one file so far (decompressed r622.udas.lfs with the BMS script above into r622.udas, then compressed r622.udas back into r622.udas.lfs with this, seemed to match exactly :D - I'll try testing it with all the games .lfs files tomorrow)

To use it just drag+drop the file you want to compress into it and it should create a .lfs file next to the input file, or you can run it like `RE4LFS.exe [path/to/input/file]`, eg. in a batch script.

Any chance you could ask Albert whether he wants to try this out? Hopefully it might help reduce the size of things a little.

**Comment by nipkownix** [comment]
Ohh! Nice work! I'll ask him later tonight. Thanks!

**Comment by emoose** [comment]
Oh there was one small issue that might have made the chunk headers different from originals, here's fixed ver:
[RE4LFS-1.0.zip](https://github.com/nipkownix/re4_tweaks/files/7551409/RE4LFS-1.0.zip)

This adds decompression support too if you use a .lfs file with it, added that because it looks like game allows LFS chunks to not be compressed, but the BMS script doesn't handle that and would probably corrupt stuff instead, so maybe this can help extract some things more properly (if there's any that use uncompressed chunks anyway)

**Comment by nipkownix** [comment]
Cool. Just sent him the updated version. I was about to suggest decompression support too! Great work!

**Comment by emoose** [comment]
Crap, seems to be some bug with compressing certain files, the .lfs it creates wont match up to the original .lfs, and decompressing that .lfs then won't match with the file that was compressed neither :/

Seems to happen with `BIO4\Em\em13.udas.lfs` and a bunch of other files, in em13.udas.lfs the final 0x1A bytes don't match the original, seems those bytes are from the last compressed block -0x8000 bytes away for some reason, looking into it more now.

**Comment by nipkownix** [comment]
> Fixed version here: RE4LFS-1.1.zip
(also fixes a bug where 1.0 wouldn't enter decompress mode if using output parameter)

Thank you! I'll send the updated version to albert. He hasn't gotten back to me yet, so he probably had no time to test it.

>Could maybe add a switch to change what byte to pad with etc, but I don't really think there's much point if the game accepts both types, doesn't look like it should refuse any at least.

Nah, no reason to bother if the game accepts both types.
How's the compression ratio, btw? I had no time to test it myself yet. I'm wondering how much of the HD project's final size can be lowered since the latest version albert sent me is quite _chonky_, and that's without the HD videos.

**Comment by emoose** [comment]
Compression ratio can be pretty good, eg. turning a 160MB 32000003.pack ImagePack into an 8MB 32000003.pack.lfs, like 5% the size 😄 
I think that was a rare case though, most of the packs usually end up at around ~28% ratio, still not too bad though.
Seems packs can also be compressed with yz2 first before being turned into LFS, maybe that could improve the ratio even more, but haven't tried it myself.

**Comment by nipkownix** [comment]
I haven't heard anything about problematic texture formats. Last time I've modded the game, it accepted both .tga and .dds files. I'll ask albert if he knows anything about that.

Also, he got back to me!

Apparently everything is working fine on his end. He also got some good compression ratio!

![image](https://user-images.githubusercontent.com/33870163/142414501-8ea40ad9-d7c2-4a4d-9fd3-8922893e7fdc.png)

**Comment by emoose** [comment]
Might be certain formats inside the DDS that it doesn't support, like DXT5/BC5/YUV etc, could be mixing it up with another game though.

Glad to hear it's working, not sure if this might be needed or not but here's a .bat file I used to handle all the .pack files in the same folder as the .bat, the .pack can be changed to any other extensions:
```
FOR %%X IN (*.pack) DO (
  C:\bin\RE4LFS.exe -f %%X
)
```

**Comment by nipkownix** [comment]
Ah, I ended up sending him the .bat I used for testing:

```
@echo off
for %%a in (uncompressed\*) do RE4LFS.exe -f "%%a" compressed\%%~nxa.lfs
```

Hopefully he can adapt it to his needs. Compressing to a different subdir is just how I like my workflow :p

Speaking about subdirs, it seems your tool fails to create the .lfs if the subdir specified doesn't exist. Could be a good idea to add code to create the subdir and also create a repo on GH for it, so it doesn't get lost to time or anything.
